Add Topic Statistics Tutorial (#824)
ros2/ros2_documentation: source/Tutorials/Topics/Topic-Statistics-Tutorial.rst

.. redirect-from::

    Topic-Statistics-Tutorial

ROS2 Topic Statistics Tutorial (C++)
====================================

**Goal:** Enable ROS2 Topic Statistics and view the output statistics data

**Tutorial level:** Beginner

**Time:** 10 minutes

.. contents:: Table of Contents
   :local:

Overview
--------

This is a short tutorial of how to enable :ref:`Topic Statistics<AboutTopicStats>`

> in ROS 2 and view the published statistics output using command line tools (:ref:`ros2topic<ROS2Topics>`).

Author: Will add a short section about Topic Statistics and call out the link to the concepts page.

Prerequisites
-------------

An installation from either binaries or source, where the version must be at least
`ROS2 Foxy <https://index.ros.org/doc/ros2/Releases/Release-Foxy-Fitzroy/>`__.
In previous tutorials, you learned how to :ref:`create a workspace <ROS2Workspace>`,
:ref:`create a package <CreatePkg>`, and create a :ref:`C++ <CppPubSub>` publisher and subscriber.

This tutorial assumes that you still have your ``cpp_pubsub`` package from the :ref:`C++ <CppPubSub>` tutorial.

Tasks
-----

1 Write the subscriber node with statistics enabled node

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Author: Typo, nice catch.

> Download the example talker code by entering the following command:

Author: Will do

        // configure the topic name (default '/statistics')
        // options.topic_stats_options.publish_topic = "/topic_statistics"

        auto callback = [this](std_msgs::msg::String::SharedPtr msg) {
            this->topic_callback(msg);
          };

        subscription_ = this->create_subscription<std_msgs::msg::String>(
          "topic", 10, callback, options);
      }

    private:
      void topic_callback(const std_msgs::msg::String::SharedPtr msg) const
      {
        RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg->data.c_str());
      }
      rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
    };

    int main(int argc, char * argv[])
    {
      rclcpp::init(argc, argv);
      rclcpp::spin(std::make_shared<MinimalSubscriberWithTopicStatistics>());
      rclcpp::shutdown();
      return 0;
    }

1.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

As in the :ref:`C++ <CppPubSub>` tutorial, we have a subscriber node which receives string messages from the
``topic`` topic from the ``topic_callback`` function.
However, what's new is we've added options to configure the subscription to enable topic statistics with

> the ``rclcpp::SubscriptionOptions()``

Author: How about options instead of function? The subscription options are a [struct](http://docs.ros2.org/foxy/api/rclcpp/structrclcpp_1_1SubscriptionOptionsBase_1_1TopicStatisticsOptions.html)

.. code-block:: C++

    // manually enable topic statistics via options
    auto options = rclcpp::SubscriptionOptions();
    options.topic_stats_options.state = rclcpp::TopicStatisticsState::Enable;

Other options, such as the statistics collection / publish period and the topic used to publish
statistics can be configured as well

.. code-block:: C++

    // configure the collection window and publish period (default 1s)
    options.topic_stats_options.publish_period = std::chrono::seconds(10);

    // configure the topic name (default '/statistics')

>     // options.topic_stats_options.publish_topic = "/my_topic"

Author: That's correct. I left the commented line so it would be easy for someone to provide their own topic name if desired.

> The configurable fields are given as

Author: How about: "The configurable fields are described in the following table:"

.. code-block:: console

  [INFO] [minimal_subscriber_with_topic_statistics]: I heard: "Hello World: 10"
  [INFO] [minimal_subscriber_with_topic_statistics]: I heard: "Hello World: 11"
  [INFO] [minimal_subscriber_with_topic_statistics]: I heard: "Hello World: 12"
  [INFO] [minimal_subscriber_with_topic_statistics]: I heard: "Hello World: 13"
  [INFO] [minimal_subscriber_with_topic_statistics]: I heard: "Hello World: 14"

Enter ``Ctrl+C`` in each terminal to stop the nodes from spinning.

3 Observe published statistic data
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

While the nodes are running, open a new terminal window.
Execute the following command:

.. code-block:: console

    ros2 topic list

This will list all currently active topics.
You should see the following:

.. code-block:: console

    /parameter_events
    /rosout
    /statistics
    /topic

If you changed the ``topic_stats_options.publish_topic`` option then you will see that instead of
``/statistics``.

The subscriber node you created is publishing statistics, for the topic ``topic``, to the output topic

> ``/statistics``.

Author: I think that's a good idea.

We can view the data published to this topic with the following command:

.. code-block:: console

    ros2 topic echo /statistics

> The terminal should start publishing statistics messages every 10 seconds, unless this duration was
> changed previously as part of the subscription configuration option ``topic_stats_options.publish_period``.

Reviewer: On line 133-134, it says that the default is 1s and it looks like you changed to 10s. But here it implies 10s is the default and that's what you'd see if you _didn't_ optionally configure it. Can you clarify?

And, in either case, I think this can be rewritten as so:

```suggestion
The terminal should start publishing statistics messages every 10 seconds, unless the 
``topic_stats_options.publish_period`` subscription configuration was optionally changed earlier in the tutorial.
```

Author: That's a good point. The default is 1 second whereas the example I gave configures it to 10 seconds (to make the statistics data appear interesting).

Reviewer: So I think (?) it'd make more sense to say:

> The terminal should start publishing statistics messages every 10 seconds, **because** the 
``topic_stats_options.publish_period`` subscription configuration was optionally changed earlier in the tutorial

Author: Sounds good.

    metrics_source: message_period
    unit: ms
    window_start:
      sec: 1594856666
      nanosec: 931527366
    window_stop:
      sec: 1594856676
      nanosec: 930797670
    statistics:
    - data_type: 1
      data: 499.2746365105009
    - data_type: 3
      data: 500.0
    - data_type: 2
      data: 499.0
    - data_type: 5
      data: 619.0
    - data_type: 4
      data: 0.4463309283488427
    ---

From the `message definition <https://github.com/ros2/rcl_interfaces/tree/master/statistics_msgs>`__
the ``data_types`` are as follows

===============    ===================
data_type value     statistics
===============    ===================
1                   average
2                   minimum
3                   maximum
4                   standard deviation
5                   sample count
===============    ===================

Here we see the two currently possible calculated statistics for the ``std_msgs::msg::String`` message published
to ``/topic`` by the ``minimal_publisher``.
Because the ``std_msgs::msg::String`` does not have a message header, the ``message_age`` calculation cannot be performed,

> so NaNs are provided.

Reviewer: Is "returned" a better word than "provided" here?

Author: Agree with returned.

topic_stats_options.state            Enable or disable topic statistics (default ``rclcpp::TopicStatisticsState::Disable``)
topic_stats_options.publish_period   The period in which to collect statistics data and publish a statistics message (default ``1s``)
topic_stats_options.publish_topic    The topic to use when publishing statistics data (default ``/statistics``)
==================================  =============================================================================================

1.2 CMakeLists.txt
~~~~~~~~~~~~~~~~~~

Now open the ``CMakeLists.txt`` file.

Add the executable and name it ``listener_with_topic_statistics`` so you can run your node using ``ros2 run``:

.. code-block:: console

    add_executable(listener_with_topic_statistics member_function_with_topic_statistics.cpp)
    ament_target_dependencies(listener_with_topic_statistics rclcpp std_msgs)

    install(TARGETS
      talker
      listener
      listener_with_topic_statistics
      DESTINATION lib/${PROJECT_NAME})

Make sure to save the file, and then your pub/sub system, with topic statistics enabled,
should be ready for use.

2 Build and run
^^^^^^^^^^^^^^^

> To build, see :ref:`cpppubsub-build-and-run`.

Reviewer: ```suggestion
To build, see the :ref:`Build and run <cpppubsub-build-and-run>` section in the pub/sub tutorial.
```
